src/Raven.Server/Documents/RetireAttachmentsSender.cs

```diff
@@ -109,185 +115,122 @@ internal async Task<int> RetireAttachments(int batchSize, long maxItemsToProcess
                                 }
 
                                 _token.ThrowIfCancellationRequested();
 
                                 if (shouldUpload == false)
                                 {
                                     // the attachment already exists in the cloud, no need to upload it again
-                                    retired.Enqueue(doc);
-                                    continue;
-                                }
-
-                                // the attachment stream is disposed by the directUpload
-                                Stream attachmentStream;
-                                if (attachmentLength.HasValue == false)
-                                {
-                                    (attachmentStream, attachmentLength) = _database.DocumentsStorage.AttachmentsStorage.GetAttachmentStreamAndLength(context, hashSlice);
-                                }
-                                else
-                                {
-                                    attachmentStream = _database.DocumentsStorage.AttachmentsStorage.GetAttachmentStream(context, hashSlice);
-                                }
-
-                                if (attachmentStream == null)
-                                {
-                                    // attachment was deleted, need to remote it from retired tree
-                                    retired.Enqueue(doc);
+                                    processed.Enqueue(doc);
                                     continue;
                                 }
 
-                                if (await directUpload.WaitForFinishedTasksIfNeededAsync(duration, _token))
-                                {
-                                    directUpload.CreateUploadTask(_database, doc, attachmentStream, hash, attachmentLength.Value);
-                                }
-                                else
-                                {
-                                    // we cannot continue the batch, need to dispose the attachment stream
-                                    await attachmentStream.DisposeAsync();
+                                var t = await Task.WhenAny(uploadTasks);
+                                uploadTasks.Remove(t);
+                                _token.ThrowIfCancellationRequested();
 
-                                    if (Logger.IsDebugEnabled)
-                                        Logger.Debug($"Timed out waiting for free slot to upload retired attachments with '{doc.LowerId}', ReadTransactionMaxOpenTimeInMs: {duration.ElapsedMilliseconds > ReadTransactionMaxOpenTimeInMs}, IsCancellationRequested: {_token.Token.IsCancellationRequested}, the upload will happen on next iteration.");
-                                }
+                                uploadTasks.Add(CreateUploadTask(directUpload, uploadedItems, doc, hash, attachmentLength));
```

Comment:
No, use a single array, and find an open slot.
Why are you not doing the _whole_ process in the `CreateUploadTask` - including the `GetObjectSizeAsync()` work?

Reply:
done